bindings/src/main/java/io/pravega/storage/hdfs/HDFSStorage.java

```diff
@@ -378,56 +408,76 @@ public void write(SegmentHandle handle, long offset, InputStream data, int lengt
             IOUtils.copyBytes(data, stream, (long) length, false);
 
             stream.flush();
+
+            HDFSMetrics.WRITE_LATENCY.reportSuccessEvent(timer.getElapsed());
+            HDFSMetrics.WRITE_BYTES.add(length);
+            LoggerHelpers.traceLeave(log, "write", traceId, handle, offset, length);
+        } catch (PathNotFoundException | FileNotFoundException e) {
+            // We reach here only if the file with current epoch is not present or sealed.
+            FileStatus status = null;
+            try {
+                status = findStatusForSegment(handle.getSegmentName(), true);
+                if (isSealed(status.getPath())) {
+                    throw new StreamSegmentSealedException(handle.getSegmentName());
+                }
+                if (getEpochFromPath(status.getPath()) != this.epoch) {
+                    throw new StorageNotPrimaryException(handle.getSegmentName());
+                }
+            } catch (IOException ex) {
+                throw HDFSExceptionHelpers.convertException(handle.getSegmentName(), ex);
+            }
         } catch (IOException ex) {
             throw HDFSExceptionHelpers.convertException(handle.getSegmentName(), ex);
         }
-
-        HDFSMetrics.WRITE_LATENCY.reportSuccessEvent(timer.getElapsed());
-        HDFSMetrics.WRITE_BYTES.add(length);
-        LoggerHelpers.traceLeave(log, "write", traceId, handle, offset, length);
     }
 
     @Override
     public SegmentHandle openWrite(String streamSegmentName) throws StreamSegmentException {
         ensureInitializedAndNotClosed();
         long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
-        long fencedCount = 0;
-        do {
+        claimOwnership(streamSegmentName);
+        LoggerHelpers.traceLeave(log, "openWrite", traceId, epoch);
+        return HDFSSegmentHandle.write(streamSegmentName);
+    }
+
+    /**
+     * Claims the ownership of this file by renaming it.
+     * To avoid problems with race condition this method retries until either rename succeeds or it discovers file with higher epoc.
+     * @param streamSegmentName
+     * @throws StreamSegmentException
+     */
+    private void claimOwnership(String streamSegmentName) throws StreamSegmentException {
+        boolean fileClaimed = false;
+        // There could be multiple containers trying to open the file.
+        // In a race, a container with lower epoch may be able to rename file before this container instance.
+        // We should try until we are able to either rename the file to epoch of current instance or find the file with higher epoch
+        while (!fileClaimed) {
```

Comment:
The previous version of this code had an upper bound on the number of attempts it would do. Even though eventually this loop should end at one point, the reason for that upper bound was to make sure we don't end up in an infinite loop from which we can never exit.

Reply:
ok. I put the upper bound back.